orchagent/dash/dashrouteorch.cpp

```diff
@@ -0,0 +1,994 @@
+extern sai_object_id_t gSwitchId;
+extern size_t gMaxBulkSize;
+
+static std::unordered_map<std::string, sai_outbound_routing_entry_action_t> sOutboundAction =
+{
+    { "vnet", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET },
+    { "vnet_direct", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_VNET_DIRECT },
+    { "route_direct", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_ROUTE_DIRECT },
+    { "drop", SAI_OUTBOUND_ROUTING_ENTRY_ACTION_DROP }
+};
+
+DashRouteOrch::DashRouteOrch(DBConnector *db, vector<string> &tableName) :
+    outbound_routing_bulker_(sai_dash_outbound_routing_api, gMaxBulkSize),
+    inbound_routing_bulker_(sai_dash_inbound_routing_api, gMaxBulkSize),
+    Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+}
+
+bool DashRouteOrch::addEniEntry(const string& eni)
+{
+    SWSS_LOG_ENTER();
+
+    if (eni_entries_[eni].eniEntryAdded)
+    {
+        SWSS_LOG_ERROR("ENI entry already exists for %s", eni.c_str());
+        return true;
+    }
+
+    EniEntry &entry = eni_entries_[eni].eni_entry;
+    const string &vnet = entry.vnet;
+
+    if (!vnet.empty() && gVnetNameToId.find(vnet) == gVnetNameToId.end())
+    {
+        SWSS_LOG_ERROR("Retry as vnet %s not found", vnet.c_str());
+        return false;
+    }
+
+    sai_object_id_t &eni_id = entry.eni_id;
+    sai_attribute_t eni_attr;
+    vector<sai_attribute_t> eni_attrs;
+
+    eni_attr.id = SAI_ENI_ATTR_VNET_ID;
+    eni_attr.value.oid = gVnetNameToId[entry.vnet];
+    eni_attrs.push_back(eni_attr);
+
+    bool has_qos = qos_entries_.find(entry.qos_name) != qos_entries_.end();
+
+    eni_attr.id = SAI_ENI_ATTR_PPS;
+    eni_attr.value.u32 = has_qos ? qos_entries_[entry.qos_name].bw : 0;
+    eni_attrs.push_back(eni_attr);
+
+    eni_attr.id = SAI_ENI_ATTR_CPS;
+    eni_attr.value.u32 = has_qos ? qos_entries_[entry.qos_name].cps : 0;
+    eni_attrs.push_back(eni_attr);
+
+    eni_attr.id = SAI_ENI_ATTR_FLOWS;
+    eni_attr.value.u32 = has_qos ? qos_entries_[entry.qos_name].flows : 0;
+    eni_attrs.push_back(eni_attr);
+
+    eni_attr.id = SAI_ENI_ATTR_ADMIN_STATE;
+    eni_attr.value.booldata = entry.admin_state;
+    eni_attrs.push_back(eni_attr);
+
+    sai_status_t status = sai_dash_eni_api->create_eni(&eni_id, gSwitchId,
+                                (uint32_t)eni_attrs.size(), eni_attrs.data());
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_ERROR("Failed to create ENI entry for %s", eni.c_str());
+        task_process_status handle_status = handleSaiCreateStatus((sai_api_t) SAI_API_DASH_ENI, status);
+        if (handle_status != task_success)
+        {
+            return parseHandleSaiStatusFailure(handle_status);
+        }
+    }
+
+    eni_entries_[eni].eniEntryAdded = true;
+    entry.eni_id = eni_id;
+    SWSS_LOG_INFO("Created ENI entry for %s", eni.c_str());
+
+    return true;
+}
+
+bool DashRouteOrch::addEniAddrMapEntry(const string& eni)
+{
+    SWSS_LOG_ENTER();
+
+    if (eni_entries_[eni].eniAddrMapEntryAdded)
```

Comment:
This might add empty entries to map for `eni` if we are not careful about [] operations. Please use `find `operation for safer checks.